src/Symfony/Component/Messenger/Adapter/PhpAmqp/Connection.php

```diff
@@ -0,0 +1,218 @@
+        $amqpOptions = array_replace_recursive(array(
+            'host' => $parsedUrl['host'] ?? 'localhost',
+            'port' => $parsedUrl['port'] ?? 5672,
+            'vhost' => isset($pathParts[0]) ? urldecode($pathParts[0]) : '/',
+            'queue' => array(
+                'name' => $queueName = $pathParts[1] ?? 'messages',
+            ),
+            'exchange' => array(
+                'name' => $queueName,
+            ),
+        ), $options);
+
+        if (isset($parsedUrl['user'])) {
+            $amqpOptions['login'] = $parsedUrl['user'];
+        }
+
+        if (isset($parsedUrl['pass'])) {
+            $amqpOptions['password'] = $parsedUrl['pass'];
+        }
+
+        if (isset($parsedUrl['query'])) {
+            parse_str($parsedUrl['query'], $parsedQuery);
+
+            $amqpOptions = array_replace_recursive($amqpOptions, $parsedQuery);
+        }
+
+        $exchangeOptions = $amqpOptions['exchange'];
+        $queueOptions = $amqpOptions['queue'];
+
+        unset($amqpOptions['queue']);
+        unset($amqpOptions['exchange']);
+
+        return new self($amqpOptions, $exchangeOptions, $queueOptions, $debug, $amqpFactory);
+    }
+
+    /**
+     * @throws \AMQPException
+     */
+    public function publish(string $body, array $headers = array())
+    {
+        if ($this->debug) {
+            $this->setup();
+        }
+
+        $this->exchange()->publish($body, null, AMQP_NOPARAM, array('headers' => $headers));
+    }
+
+    /**
+     * Waits and gets a message from the configured queue.
+     *
+     * @throws \AMQPException
+     */
+    public function waitAndGet(): ?\AMQPEnvelope
+    {
+        if ($this->debug) {
+            $this->setup();
+        }
+
+        $message = null;
+
+        try {
+            $this->queue()->consume(function (\AMQPEnvelope $envelope) use (&$message): bool {
+                $message = $envelope;
+
+                return false;
+            });
+        } catch (\AMQPQueueException $e) {
+            if (404 === $e->getCode()) {
+                // If we get a 404 for the queue, it means we need to setup the exchange & queue.
+                $this->setup();
+
+                return $this->waitAndGet();
+            }
+
+            throw $e;
+        }
+
+        return $message;
+    }
+
+    public function ack(\AMQPEnvelope $message)
+    {
+        return $this->queue()->ack($message->getDeliveryTag());
+    }
+
+    public function reject(\AMQPEnvelope $message)
+    {
+        return $this->queue()->reject($message->getDeliveryTag());
+    }
+
+    public function nack(\AMQPEnvelope $message)
+    {
+        return $this->queue()->nack($message->getDeliveryTag());
+    }
+
+    public function setup()
+    {
+        if (!$this->channel()->isConnected()) {
+            $this->clear();
+        }
+
+        $this->exchange()->declareExchange();
+
+        $this->queue()->declareQueue();
+        $this->queue()->bind($this->exchange()->getName());
```

Comment:
This works because the default exchange type is Fan Out. If one use Direct is will not work anymore. You may add a routing key option to configure it

Reply:
Good point, added the routing key as a possible option on the queue. `queue[routing_key]=...` in the DSN 👍